File: src/main/java/io/diekema/dingo/feast/sources/SimpleFileSystemSource.java
/*
 * Copyright (c) 2016. Richard Diekema - All Rights Reserved
 * Unauthorized copying of this file, via any medium is strictly prohibited Proprietary and confidential
 * Written by Richard Diekema <rdiekema@gmail.com>, September 2016
 */

package io.diekema.dingo.feast.sources;

import io.diekema.dingo.feast.Asset;
import io.diekema.dingo.feast.Exchange;
import io.diekema.dingo.feast.Source;
import org.apache.commons.io.IOUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.InputStream;
import java.nio.file.*;
import java.nio.file.attribute.BasicFileAttributes;
import java.util.Collections;
import java.util.LinkedList;
import java.util.List;
import java.util.concurrent.atomic.AtomicInteger;

import static java.nio.charset.StandardCharsets.UTF_8;

/**
 * Created by rdiekema on 9/25/16.
 */
public class SimpleFileSystemSource implements Source {

    private static final Logger log = LoggerFactory.getLogger(RecursiveFileSystemSource.class.getName());

    private String inputPath;
    private String inputPatternAndSyntax;

    public SimpleFileSystemSource(String filePath, String syntaxAndPattern) {
        this.inputPath = filePath;
        this.inputPatternAndSyntax = syntaxAndPattern;
    }

    @Override
    public Exchange collect() throws IOException {
        final PathMatcher pathMatcher = FileSystems.getDefault().getPathMatcher(inputPatternAndSyntax);
        final AtomicInteger fileCount = new AtomicInteger(0);
        final List<FilePayload> payloads = new LinkedList<>();

        Files.walkFileTree(FileSystems.getDefault().getPath(inputPath), Collections.<FileVisitOption>emptySet(), 1, new SimpleFileVisitor<Path>() {
            @Override
            public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IOException {
                if (pathMatcher.matches(file)) {
                    fileCount.incrementAndGet();
                    payloads.add(new FilePayload(file));
                }

                return FileVisitResult.CONTINUE;
            }

            @Override
            public FileVisitResult visitFileFailed(Path file, IOException exc) throws IOException {
                return FileVisitResult.CONTINUE;
            }
        });

        log.info(String.format("Found %d files.", fileCount.intValue()));

        List<Asset> assets = new LinkedList<>();
        for (FilePayload filePayload : payloads) {
            assets.add(new Asset(filePayload.getName(), IOUtils.toString(filePayload.getInputStream(), UTF_8), filePayload.getPath()));
        }

        Exchange exchange = new Exchange();
        exchange.enrich(assets);

        return exchange;
    }

    static class FilePayload {
        String name;
        Path path;
        InputStream inputStream;

        public FilePayload(Path file) throws FileNotFoundException {
            this.path = file;
            this.name = file.getFileName().toString();
            this.inputStream = new FileInputStream(file.toFile());
        }

        public String getName() {
            return name;
        }

        public Path getPath() {
            return path;
        }

        public InputStream getInputStream() {
            return inputStream;
        }
    }

}

File: src/main/kotlin/io/diekema/dingo/feast/sources/SimpleFileSystemSource.kt
/*
 * Copyright (c) 2016. Richard Diekema - All Rights Reserved
 * Unauthorized copying of this file, via any medium is strictly prohibited Proprietary and confidential
 * Written by Richard Diekema <rdiekema@gmail.com>, September 2016
 */

package io.diekema.dingo.feast.sources

import io.diekema.dingo.feast.Asset
import io.diekema.dingo.feast.Exchange
import org.apache.commons.io.IOUtils
import org.slf4j.Logger
import org.slf4j.LoggerFactory

import java.io.FileInputStream
import java.io.FileNotFoundException
import java.io.IOException
import java.io.InputStream
import java.nio.file.*
import java.nio.file.attribute.BasicFileAttributes
import java.util.Collections
import java.util.LinkedList
import java.util.concurrent.atomic.AtomicInteger

import java.nio.charset.StandardCharsets.UTF_8

/**
 * Created by rdiekema on 9/25/16.
 */
class SimpleFileSystemSource(private val inputPath: String, private val inputPatternAndSyntax: String) : Source {

    @Throws(IOException::class)
    override fun collect(): Exchange {
        val pathMatcher = FileSystems.getDefault().getPathMatcher(inputPatternAndSyntax)
        val fileCount = AtomicInteger(0)
        val payloads = LinkedList<FilePayload>()

        Files.walkFileTree(FileSystems.getDefault().getPath(inputPath), emptySet<FileVisitOption>(), 1, object : SimpleFileVisitor<Path>() {
            @Throws(IOException::class)
            override fun visitFile(file: Path, attrs: BasicFileAttributes): FileVisitResult {
                if (pathMatcher.matches(file)) {
                    fileCount.incrementAndGet()
                    payloads.add(FilePayload(file))
                }

                return FileVisitResult.CONTINUE
            }

            @Throws(IOException::class)
            override fun visitFileFailed(file: Path, exc: IOException?): FileVisitResult {
                return FileVisitResult.CONTINUE
            }
        })

        log.info(String.format("Found %d files.", fileCount.toInt()))

        val assets = LinkedList<Asset>()
        for (filePayload in payloads) {
            assets.add(Asset(filePayload.name, IOUtils.toString(filePayload.inputStream, UTF_8), filePayload.path.toString()))
        }

        val exchange = Exchange()
        exchange.enrich(assets)

        return exchange
    }

    internal class FilePayload @Throws(FileNotFoundException::class)
    constructor(var path: Path) {
        var name: String
        var inputStream: InputStream

        init {
            this.name = path.fileName.toString()
            this.inputStream = FileInputStream(path.toFile())
        }
    }

    companion object {

        private val log = LoggerFactory.getLogger(SimpleFileSystemSource::class.java.name)
    }

}
